### ISSUE TITLE
Hercules has crashed

### ISSUE THREAD

**Issue by werner02815** [SDL-Hercules-390/hyperion#187]
I am running zOS system on "Hercules version 4.1.0.9426-SDL-g42b533fa (4.1.0.9426)".

Yesterday hercules crashed 2 times. (It has always run stable before and after then!) Unfortunately I am not able to recreate problem at will.

What I have seen however (which may be a clue), is that each time the problem occurred was when I was running a `ICKDSF` job to format a 3390-54 dasd volume. However, doing same procedure again seems to work ok.

The Hercules.dmp for the 2nd occurence is available, but is too large to upload here. Can you tell where to send it? Please also let me know any additional information you need, so I can package it all together.

Thanks in advance.

Here's a screen shot of the HerGUI log file from one of the crashes:

![grafik](https://user-images.githubusercontent.com/47626430/52775205-66f64800-303f-11e9-891a-75aa66775932.png)

**Comment by werner02815** [comment]
now have located some "normal" work which causes hercules to crash with a 50% opportunity. So in other words i am able to recreate based on any requests. I have already tried different configurations with parameters DEVTMAX (changed from 0 to 25, no effect), removed any CCKD parameters (CCKD  COMPPARM=9,NOSFD=1 removed, no effect). 
I never used model larger than 3390-3 in past, but now i am required to do so. I will try to find a bypass to still using 3390-3 and check whether this is root cause. Giving feedback tomorrow.

**Comment by werner02815** [comment]
additional analysis done with following results: problem is based on 3390 model type. Running procedure with 3390-3 or 3390-9 works fine. Running same procedure with 3390-27 or 3390-54 crashes Hercules (running with 3390-54 it seems that problem occurs "more often"). 

DASD volumes are created with following command:
dasdinit.exe 	-z 	U:\PSI_DUMP\H5SM00 	3390-54		H5SM00

following config is used:
ARCHLVL              z/Arch
MAXCPU               2
ENGINES              CP,CP
NUMCPU               2
AUTO_SCSI_MOUNT      NO
CCKD                 COMPPARM=9,NOSFD=1
CNSLPORT             3270
CONKPALV             (3,1,10)
CODEPAGE             default
CPUMODEL             3906
CPUSERIAL            012345
CPUVERID             84
DEVTMAX              0
DIAG8CMD             ENABLE ECHO
ECPSVM               NO
HTTP                 PORT 80 NOAUTH
MODPATH              .
LEGACYSENSEID        DISABLE
LOADPARM             30022DM1
LPARNUM              01
LPARNAME             LPAR1
MAINSIZE             1024
MANUFACTURER         IBM
MODEL                M02 501 502 1234
MOUNTED_TAPE_REINIT  DISALLOW
NETDEV               02-00-4C-4F-4F-50
OSTAILOR             QUIET
PANRATE              50
PGMPRDOS             LICENSED
PLANT                84
SHCMDOPT             NODIAG8
SYSEPOCH             1900
TIMERINT             50
TRACEOPT             TRADITIONAL
TZOFFSET             +0100
XPNDSIZE             0
YROFFSET             0

3100    3390    U:/PSI_DUMP/H5SM00 cu=3990-6 

operating system is zOS 1.13, so it really should support 3390-54

**Comment by Fish-Git** [comment]
Given that you are using large model dasd (i.e. dasd models that are natively much larger than 4GB; 3390-27 are 27GB and 3390-54 are 55GB) and the fact you have created them as compressed "CCKD" dasd images, I highly suspect your problem is that your dasd image files are exceeding the maximum file size that the CCKD format can support (which is only 4GB).

To confirm or deny this suspicion, please show us the output of a `dir` command that shows us the actual size of your `U:/PSI_DUMP/H5SM00` dasd image file, e.g:

```
dir U:/PSI_DUMP
```

If the size of your H5SM00 dasd image file is very close to 4GB, then that's your problem. Once the size of the file reaches or exceeds 4GB (which can _easily_ occur when 3390-27 or 3390-54 model dasd is used!), strange things will happen (with the mostly likely "strange thing" being Hercules crashing).

Until version 4.2 is officially released (which will contain CCKD64 support), I'm afraid you'll have to use regular _non-compressed_ CKD dasds instead. (I.e. when you do your `dasdinit` to create your dasd image file, do _not_ specify the `-z` option). Be aware that this of course means your dasd image files will be huge (27GB or 55GB), so make sure you have plenty of disk space beforehand.

**Comment by Fish-Git** [comment]
Please _compress_ your "Hercules.dmp" crash dump file (using e.g. WinZip or some other file compression tool) and upload it via FTP to:

   ftp://www.softdevlabs.com/incoming/

Thanks.

**Comment by werner02815** [comment]
I see crash even when i first initialize volume within zOS guest system, so it is no issue of 4 GB limit (nevertheless thanks for pointing to this limit). I have dumps from various situations, each time DASD image is far away from this 4 GB limit. Unfortunately i am not so familiar with this git upload. What user/pwd has to be used for ftp`ing ? Using anonymous ftp i get "553 Could not create file.", trying to login with GIT account i get "The PASS command was not accepted."

**Comment by Fish-Git** [comment]
> Unfortunately I am not so familiar with this git upload.

I'm not asking you to upload it to githib.

I am asking you to upload it _to my own server._&nbsp;(*)

(*) ([softdevlabs](http://www.softdevlabs.com) is my own private domain where I have a web, ftp and mail server running.)

> What user/pwd has to be used for ftp'ing?

[Standard anonymous FTP](https://en.wikipedia.org/wiki/File_Transfer_Protocol#Anonymous_FTP): user name: anonymous, password: &lt;your email address&gt;.

> Using anonymous ftp I get "553 Could not create file."

While I personally prefer using [FileZilla](https://en.wikipedia.org/wiki/FileZilla) myself, you can use pretty much _any_ FTP client you want, even the one that comes with Windows:

```dos.bat
  Microsoft Windows [Version 6.1.7601]
  Copyright (c) 2009 Microsoft Corporation.  All rights reserved.

  C:\Users\Fish>which ftp

    "C:\Windows\System32\ftp.exe"

  C:\Users\Fish>ftp
  ftp> open www.softdevlabs.com
  Connected to softdevlabs.com.
  220 (vsFTPd 2.2.2)
  User (softdevlabs.com:(none)): anonymous
  230 Login successful.
  ftp> cd incoming
  250 Directory successfully changed.
  ftp> ?
  Commands may be abbreviated.  Commands are:

  !               delete          literal         prompt          send
  ?               debug           ls              put             status
  append          dir             mdelete         pwd             trace
  ascii           disconnect      mdir            quit            type
  bell            get             mget            quote           user
  binary          glob            mkdir           recv            verbose
  bye             hash            mls             remotehelp
  cd              help            mput            rename
  close           lcd             open            rmdir
  ftp> binary
  200 Switching to Binary mode.
  ftp> put fishtest.txt
  200 PORT command successful. Consider using PASV.
  150 Ok to send data.
  226 Transfer complete.
  ftp: 10 bytes sent in 0.11Seconds 0.09Kbytes/sec.
  ftp> bye
  221 Goodbye.

  C:\Users\Fish>
```

**Comment by werner02815** [comment]
sorry, i forget on cd incoming ...

i have packaged now 3 different dumps and in addition herc.conf and uploaded it with name werner_hercdump.zip. It`s size is 231760975 bytes

**Comment by Fish-Git** [comment]
> I have packaged now 3 different dumps and in addition herc.conf and uploaded it with name werner_hercdump.zip. It's size is 231760975 bytes.

Got it! Thanks. I'll take a look at it and let you know what I find (if anything).

**Comment by Fish-Git** [comment]
I have analyzed each dump and they're all the same. They're all crashing in the exact same spot: in the CCKD Garbage Collector "relocate level 2 tables" function (`cckd_gc_l2`):
___
```
Microsoft (R) Windows Debugger Version 6.12.0002.633 AMD64
Copyright (c) Microsoft Corporation. All rights reserved.

Loading Dump File [C:\Users\Fish\Documents\Visual Studio 2008\Projects\Hercules\_DUMPS\Hercules.dmp3]
User Mini Dump File: Only registers, stack and portions of memory are available

WARNING: Minidump contains unknown stream type 0x15
WARNING: Minidump contains unknown stream type 0x16
Comment: '4.1.0.9426-SDL-g42b533fa'
Comment: '(C) Copyright 1999-2018 by Roger Bowler, Jan Jaeger, and others'
Comment: 'Running on: H01905429 (Windows-6.2.9200 Intel(R) x64) LP=4, Cores=2, CPUs=1'
Comment: 'Built with: Microsoft Visual Studio 2008 (MSVC 150030729 1)'
Comment: 'Build type: Windows MSVC AMD64 host architecture build'
Comment: 'Modes: S/370 ESA/390 z/Arch'
Comment: 'Max CPU Engines: 64'
Comment: 'Using   shared libraries'
Comment: 'Using   Fish threads Threading Model'
Comment: 'Using   Error-Checking Mutex Locking Model'
Comment: 'With    Shared Devices support'
Comment: 'With    Dynamic loading support'
Comment: 'With    External GUI support'
Comment: 'With    Partial TCP keepalive support'
Comment: 'With    IPV6 support'
Comment: 'With    HTTP Server support'
Comment: 'With    sqrtl support'
Comment: 'Without SIGABEND handler'
Comment: 'With    CCKD BZIP2 support'
Comment: 'With    HET BZIP2 support'
Comment: 'With    ZLIB support'
Comment: 'With    Regular Expressions support'
Comment: 'With    Object REXX support'
Comment: 'Without Regina REXX support'
Comment: 'With    Automatic Operator support'
Comment: 'Without National Language Support'
Comment: 'Machine dependent assists: cmpxchg1 cmpxchg4 cmpxchg8 hatomics=msvcIntrinsics'
Symbol search path is: C:\Users\Fish\Documents\Visual Studio 2008\Projects\Hercules\_DUMPS;srv*C:\Windows\Symbols*http://msdl.microsoft.com/download/symbols
Executable search path is: C:\Users\Fish\Documents\Visual Studio 2008\Projects\Hercules\_DUMPS
Windows 7 Version 16299 MP (4 procs) Free x64
Product: WinNt, suite: SingleUserTS
Machine Name:
Debug session time: Thu Feb 14 03:32:49.000 2019 (UTC - 8:00)
System Uptime: not available
Process Uptime: 0 days 1:16:01.000
................................................................
...........................................
Loading unloaded module list
..
This dump file has an exception of interest stored in it.
The stored exception information can be accessed via .ecxr.
(15f4.337c): Access violation - code c0000005 (first/second chance not available)
ntdll!ZwGetContextThread+0x14:
00007ff8`373d1e04 c3              ret
0:012> !analyze -v
*******************************************************************************
*                                                                             *
*                        Exception Analysis                                   *
*                                                                             *
*******************************************************************************

Unable to load image C:\Program Files (x86)\SoftDevLabs\CTCI-WIN\TunTap64.dll, Win32 error 0n2
*** WARNING: Unable to verify timestamp for TunTap64.dll
Failed calling InternetOpenUrl, GLE=12029

FAULTING_IP: 
hdasd!cckd_gc_l2+340 [c:\users\fish\documents\visual studio 2008\projects\hercules\_git\_fish\hyperion-0\cckddasd.c @ 5068]
00000000`03b30140 418b08          mov     ecx,dword ptr [r8]

EXCEPTION_RECORD:  ffffffffffffffff -- (.exr 0xffffffffffffffff)
ExceptionAddress: 0000000003b30140 (hdasd!cckd_gc_l2+0x0000000000000340)
   ExceptionCode: c0000005 (Access violation)
  ExceptionFlags: 00000000
NumberParameters: 2
   Parameter[0]: 0000000000000000
   Parameter[1]: 0000000008a63010
Attempt to read from address 0000000008a63010

DEFAULT_BUCKET_ID:  INVALID_POINTER_READ

PROCESS_NAME:  hercules.exe

ERROR_CODE: (NTSTATUS) 0xc0000005 - The instruction at 0x%08lx referenced memory at 0x%08lx. The memory could not be %s.

EXCEPTION_CODE: (NTSTATUS) 0xc0000005 - The instruction at 0x%08lx referenced memory at 0x%08lx. The memory could not be %s.

EXCEPTION_PARAMETER1:  0000000000000000

EXCEPTION_PARAMETER2:  0000000008a63010

READ_ADDRESS:  0000000008a63010 

FOLLOWUP_IP: 
hdasd!cckd_gc_l2+340 [c:\users\fish\documents\visual studio 2008\projects\hercules\_git\_fish\hyperion-0\cckddasd.c @ 5068]
00000000`03b30140 418b08          mov     ecx,dword ptr [r8]

MOD_LIST: <ANALYSIS/>

NTGLOBALFLAG:  0

APPLICATION_VERIFIER_FLAGS:  0

FAULTING_THREAD:  000000000000337c

PRIMARY_PROBLEM_CLASS:  INVALID_POINTER_READ

BUGCHECK_STR:  APPLICATION_FAULT_INVALID_POINTER_READ

LAST_CONTROL_TRANSFER:  from 0000000003b2f689 to 0000000003b30140

STACK_TEXT:  
00000000`0b56f760 00000000`03b2f689 : 00000000`080a7520 00000000`00000000 00000000`00040000 00000000`080a7510 : hdasd!cckd_gc_l2+0x340 [c:\users\fish\documents\visual studio 2008\projects\hercules\_git\_fish\hyperion-0\cckddasd.c @ 5068]
00000000`0b56f7c0 00000000`03b2f1dd : 00000000`00000100 00000000`080c6000 00000000`00000000 00000000`00000100 : hdasd!cckd_gc_percolate+0x159 [c:\users\fish\documents\visual studio 2008\projects\hercules\_git\_fish\hyperion-0\cckddasd.c @ 4775]
00000000`0b5af8f0 00000000`034bff99 : 00000000`0000337c 00010498`5c6551cc 00000000`5c6551cc 00000000`705b334a : hdasd!cckd_gcol+0x52d [c:\users\fish\documents\visual studio 2008\projects\hercules\_git\_fish\hyperion-0\cckddasd.c @ 4681]
00000000`0b5af9c0 00000000`034b3bf4 : 00000000`0a17bf10 00000000`00000000 00000000`00000000 00000000`00000000 : hutil!hthread_func+0x49 [c:\users\fish\documents\visual studio 2008\projects\hercules\_git\_fish\hyperion-0\hthreads.c @ 797]
00000000`0b5af9f0 00000000`705b2fdf : 00000000`7063d790 00000000`00000000 00000000`00000000 00000000`00000000 : hutil!FTWin32ThreadFunc+0x44 [c:\users\fish\documents\visual studio 2008\projects\hercules\_git\_fish\hyperion-0\fthreads.c @ 784]
00000000`0b5afa40 00000000`705b3080 : 00000000`7063cad0 00000000`0829c440 00000000`00000000 00000000`00000000 : msvcr90!_callthreadstartex+0x17 [f:\dd\vctools\crt_bld\self_64_amd64\crt\src\threadex.c @ 348]
00000000`0b5afa70 00007ff8`36231fe4 : 00000000`00000000 00000000`00000000 00000000`00000000 00000000`00000000 : msvcr90!_threadstartex+0x84 [f:\dd\vctools\crt_bld\self_64_amd64\crt\src\threadex.c @ 326]
00000000`0b5afaa0 00007ff8`3739cb81 : 00000000`00000000 00000000`00000000 00000000`00000000 00000000`00000000 : kernel32!BaseThreadInitThunk+0x14
00000000`0b5afad0 00000000`00000000 : 00000000`00000000 00000000`00000000 00000000`00000000 00000000`00000000 : ntdll!RtlUserThreadStart+0x21

FAULTING_SOURCE_CODE:  
  5064:             )
  5065:            )
  5066:         {
  5067:             for (i = 0; i < cckd->cdevhdr[sfx].num_L1tab; i++)
> 5068:                 if (fpos + cckd->ifb[i].ifb_len == (off_t)cckd->L1tab[sfx][i])
  5069:                     break;
  5070:         }
  5071:         else
  5072:         {
  5073:             for (i = 0; i < cckd->cdevhdr[sfx].num_L1tab; i++)

SYMBOL_STACK_INDEX:  0

SYMBOL_NAME:  hdasd!cckd_gc_l2+340

FOLLOWUP_NAME:  MachineOwner

MODULE_NAME: hdasd

IMAGE_NAME:  hdasd.dll

DEBUG_FLR_IMAGE_TIMESTAMP:  5be725cd

STACK_COMMAND:  ~12s; .ecxr ; kb

FAILURE_BUCKET_ID:  INVALID_POINTER_READ_c0000005_hdasd.dll!cckd_gc_l2

BUCKET_ID:  X64_APPLICATION_FAULT_INVALID_POINTER_READ_hdasd!cckd_gc_l2+340

WATSON_STAGEONE_URL:  http://watson.microsoft.com/StageOne/hercules_exe/4_1_0_9426/5be72602/hdasd_dll/4_1_0_9426/5be725cd/c0000005/00010140.htm?Retriage=1

Followup: MachineOwner
---------
```
___
&nbsp;
Once again this smells _very_ much like a CCKD 4GB limit bug.

> I see crash even when I first initialize volume within zOS guest system, so it is no issue of 4 GB limit (nevertheless thanks for pointing to this limit). I have dumps from various situations, each time DASD image is far away from this 4 GB limit.

I understand that.  But due to the way CCKD support is written (designed), the fact that the image file -- as seen by the host -- is still much less than 4GB is actually not all that surprising.

Hercules's CCKD support, in an effort to be as efficient as possible, does a lot of internal caching and doesn't actually flush any of its changes to the actual physical file (as seen by the host) until much later. Until then, for efficiency (and for file integrity purposes as well), it just maintains them in core (in memory) and simply marks them "pending", to be eventually written out whenever it has the time. (And if you're flooding it with ICKDSF work, it might not be able to find the time!)

You are doing `ICKDSF` on an _empty_ 3390-54 (or 3390-27, etc) dasd, which causes a lot of updates to the volume (causes a lot of data to be written, a lot of tracks to be updated), and each update requires free space, which needs to come from somewhere.

When space is needed, it is acquired from the end of the file (since, being a brand new empty CCKD image, that's the only place where free space exists!), causing the location of the _next_ available free space to move further and further, until eventually the location of the level 2 tables and track images for a given track (which came from wherever the free space existed at the time) are beyond the 4GB internal limit, causing all kinds of weird problems to occur due to the internal pointers being only 32-bits wide.

At least that's my current theory anyway (i.e. best guess).&nbsp;`:)`

_**QUESTION:**_ what ICKDSF parameters are you using? Can you provide us with the ICKDSF jobstream you are running on z/OS? I would like to test my theory on my own system to see if I can recreate your crash, but I am inexperienced with z/OS. I've never run a ICKDSF under z/OS before. Can you help me out? Can you provide to me what your ICKDSF JCL looks like? Thanks!

**Comment by werner02815** [comment]
i am using following procedure to format a DASD volume to z/OS:
//OFFLINE  EXEC PGM=SYSCMD                                              
//SYSIN    DD *                                                         
V 2100,OFFLINE                                                          
WAIT=3                                                                  
//*                                                                     
//INIT     EXEC PGM=ICKDSF,TIME=NOLIMIT,REGION=3M,PARM='NOREPLYU'       
//SYSPRINT DD SYSOUT=*                                                  
//SYSIN    DD *                                                         
    INIT UNITADDRESS(2100) VOLID(TEST00) VERIFY(TEST00) -               
    NODSEXIST NOCHECK NOVALIDATE PURGE MAP NORECLAIM OWNER(SYS1) -      
    SG -                                                                
    VTOC(0,1,299) INDEX(20,0,45)                                        
//*                                                                     
//ONLINE   EXEC PGM=SYSCMD                                              
//SYSIN    DD *                                                         
V 2100,ONLINE                                                           
//*                                                                     
//VVDS     EXEC PGM=IDCAMS                                              
//SYSPRINT DD SYSOUT=*                                                  
//SYSIN    DD *                                                         
 DEF CL(NAME(SYS1.VVDS.VTEST00) VOL(TEST00) STORAGECLASS(NONSMS) -      
     NONINDEXED CYLINDERS(5,1))                                         
//*                                                                     

Today i tried to further analyze and rerun this format job again many hundred times and was not able to recreate. Absolutely frustrating ! What i have changed is placement of dasd files to another disk subsystem because of space constraint. I will try to relocate again tomorrow and try to find a way to more accurate to recreate.

**Comment by werner02815** [comment]
I have done a lot of additional tests but i do not find a procedure on how to "force" this error condition. Nevertheless it is still coming up. Nevertheless i have setup device without compression and i can confirm that in this config i do not see this crash. So your idea is most likely valid.

Unfortunately working with all uncompressed volumes is not really a solution for me. You mentioned about CCKD64 support. I am willing to do any beta test, but how to get a prebuild version including CCKD64 support ?

**Comment by Fish-Git** [comment]
> Is it possible to obtain a prebuilt version of SDL Hyperion 4,2 with CCKD64 support so I can try it without having to build it for myself?

I would prefer to not publish an interim release of SDL Hyperion 4.2 since it is still under development. I would prefer to wait until several more pending issues are resolved first, and then do an official release.

_**However, ...**_ since you are not the first person to make this request, _and_ since CCKD64 support is such an important new feature (and the more people that test it the better), I will see what I can do.

Give me a few days (hopefully it will not take longer than that) and I will try to publish a new pre-built snapshot for you and others to test with.

I will let you know when it is ready.

**Comment by Fish-Git** [comment]
> I will let you know when it is ready.

@werner02815: &nbsp; _**FYI:** &nbsp;It is now ready._

A _**temporary**_ pre-built of SDL Hyperion 4.2  is now available for download via the following download link:

* [SDL_TEMP_4.2_PVT_X64_VS2017.zip](http://www.softdevlabs.com/downloads/SDL_TEMP_4.2_PVT_X64_VS2017/SDL_TEMP_4.2_PVT_X64_VS2017.zip)

_**PLEASE NOTE**_ that it will _**not**_ remain there forever. It will disappear once 4.2 is officially released. This is only an _interim_ pre-built to allow people such as yourself to test the new CCKD64 logic.

Please create a _new_ (separate) GitHub Issue for any problem(s) you may have with it. _(which will hopefully be none!)_

In the mean time, I am going to go ahead and close this issue at this time since I strongly suspect the problem is very likely a CCKD 32-bit file format file size limitation bug.(*)
___
(*) I cannot prove that suspicion right now, but attempting to do so would probably waste more time than it would be worth. Each time I've seen this bug it has always been with people trying to load data onto a compressed (CCKD) 3390 dasd whose _native_ file size is much greater than 4GB.

**Comment by Fish-Git** [comment]
Closed _(forever!)_ by commit 6e4c52cda8c1354a658263beb2aac8b343b01252.
